### PR TITLE
Add inplace upgrader controller to hcco

### PR DESCRIPTION
**Pull request by enxebre**
**What this PR does / why we need it**:
This PR distributes the inplace upgrade logic from the NodePool controller into individual controllers running for each control plane namespace.
This let us honour our intentional network design restrictions to access hosted clusters and scope the impact and resources for each controller to specific hosted clusters.
The implementation at the moment uses annotations on MachineSets to communicate upgrade progress between NodePool controller (NodePool <-> MachineSet) and inplace upgrader (MachineSet <-> Nodes) i.e NodePool <-> Nodes. This info is modeled in the nodePoolUpgradeAPI struct which in future might be exposed as a CRD if needed.
The controller is added to the hcco manager as it already solves the problem of watching and caching targeted resources in both management and hosted cluster.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Ref https://issues.redhat.com/browse/HOSTEDCP-427.

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [x] This change includes unit tests.

### CODE REVIEW

**Comment by enxebre** [comment]
/hold
until we enable NodePool upgrade e2e tests to be required on CI.

**Review by yuqi-zhang — COMMENTED** [review]
Did an initial pass over the code. The general direction makes sense. I am not very familiar with the hostedclusteroperator, so just wanted to check my understanding that:

1. The nodepool_controller now only reconciles the machineset
2. The new inplaceupgrader runs as a new controller in the hostedclusteroperator pod (running on the control plane of the hosted cluster) that is able to read both management cluster and hosted cluster objects, and does the actual reconciling
3. The two above operate independently of each other? Is there any inter-dependency?

Also probably should hold on https://github.com/openshift/hypershift/pull/1335 until this one merges, and then switch this over

**Comment by enxebre** [comment]
Needed https://github.com/openshift/release/pull/28623 to run latest CPO
/test e2e-aws-all

**Comment by enxebre** [comment]
```
    --- FAIL: TestInPlaceUpgradeNodePool/EnsureAPIBudget (0.38s)
        --- PASS: TestInPlaceUpgradeNodePool/EnsureAPIBudget/control-plane-operator_read (0.14s)
        --- FAIL: TestInPlaceUpgradeNodePool/EnsureAPIBudget/control-plane-operator_mutate (0.04s)
        --- PASS: TestInPlaceUpgradeNodePool/EnsureAPIBudget/control-plane-operator_no_404_deletes (0.02s)
        --- PASS: TestInPlaceUpgradeNodePool/EnsureAPIBudget/ignition-server_p90_payload_generation_time (0.02s)
        --- PASS: TestInPlaceUpgradeNodePool/EnsureAPIBudget/hypershift-operator_read (0.02s)
        --- PASS: TestInPlaceUpgradeNodePool/EnsureAPIBudget/hypershift-operator_mutate (0.02s)
        --- PASS: TestInPlaceUpgradeNodePool/EnsureAPIBudget/hypershift-operator_no_404_deletes (0.02s)
```

Increasing inplace-upgrader-controller Increase control-plane-operator mutate budget.

/test e2e-aws-all

**Comment by enxebre** [comment]
```=== RUN   TestInPlaceUpgradeNodePool/EnsureAPIBudget/hypershift-operator_mutate
    util.go:548: over budget: budget: 6000, actual: 6134
```

/test e2e-aws-all

**Comment by enxebre** [comment]
temporary disabling replace upgrade as it's timing out to analyse api budgets 

/test e2e-aws-all

**Comment by enxebre** [comment]
```
  * could not run steps: step e2e-aws-all failed: "e2e-aws-all" pre steps failed: "e2e-aws-all" pod "e2e-aws-all-ipi-install-rbac" failed: the pod ci-op-rsb80nt5/e2e-aws-all-ipi-install-rbac failed after 39s (failed containers: test): ContainerFailed one or more containers exited
```
Random Infra failure

/test e2e-aws-all

**Comment by enxebre** [comment]
PTAL @alvaroaleman @csrwng

**Review by alvaroaleman — APPROVED** [review]

**Review by yuqi-zhang — APPROVED** [review]
Did some manual testing with machineconfigs, functionality lgtm. Thanks for the patch!

**Comment by enxebre** [comment]
https://github.com/openshift/hypershift/pull/1399 e2e-aws-all prevents e2e from consistently timing out.
So I'm dropping the commit that disabled the replace upgrade test from this PR.
Both in place and replace should consistently pass feature validations now. If budgeting fails lets address separately.

/e2e-aws-all

**Review by davidvossel — COMMENTED** [review]
@enxebre  If we're making a new controller entirely to handle inplace updates, was making the controller reconcile machines directly instead of MachineSets ever considered?

it's kind of awkward working with machine sets and then having to iterate over all the machines/nodes. I imagine that will get more awkward once drain is implemented.

Is there a reason why we couldn't reconcile machines directly and compare the version of the node to the version in the corresponding nodepool in order to determine if an inplace update of the machine is necessary?

**Comment by yuqi-zhang** [comment]
> Is there a reason why we couldn't reconcile machines directly and compare the version of the node to the version in the corresponding nodepool in order to determine if an inplace update of the machine is necessary?

So from my loose understanding, the issue here is that the upgrade path and the MachineSet path isn't directly the same path. MachineSets are used for scaling new nodes. Upgrade controller is used to create MachineConfig related artifacts to upgrade the node via a special MachineConfigDaemon mode, which don't use Machines.

Put another way, if you create a machineconfig (via a configmap) in the management cluster to create a new systemd service, and associate it with a hostedcluster so it rolls out to a nodepool of workers, I don't think the Machines reflect that change at all.

**Comment by yuqi-zhang** [comment]
> The MachineSets don't appear to directly reflect this change either. From what I see the inplace upgrade is triggered by the nodepool controller by writing an annotation to the machineset.

Right, sorry, I think I misunderstood your initial concern.

To the point why we integrated MachineSets into this flow, I think it's because we just decided to go a step down from the MachineDeployments that the replace upgrade uses, while keeping around some features. @enxebre definitely would know better.

I don't know if this relevant for Hypershift but e.g. for self-driving OCP, we have workflows that e.g. allow you to partition disks differently if you opt to use different instance types (different device naming) for your worker pool. This is done via a secret which we create a new machineset to use. Having MachineSets may be useful for scenarios like that eventually?

**Comment by davidvossel** [comment]
>To the point why we integrated MachineSets into this flow, I think it's because we just decided to go a step down from the MachineDeployments that the replace upgrade uses, while keeping around some features. @enxebre definitely would know better.

to clarify, I wasn't questioning the usage of MachineSets by the NodePool controller for inplace. That makes sense.

I was questioning whether it might reduce complexity to have the InPlaceUpdate controller work on Machines directly rather than reconciling MachineSets (which results in a bunch of looping logic to handle machine/node lists)

Back when the InPlace update was being handled by the NodePool controller, it made sense to work on the MachineSet because that was a 1:1 relationship to the NodePool, but now that we have an entirely dedicated controller dedicated just for inplace, it seems like the Machine is the entity to work on and not the MachineSet.

**Pull request merged**
